theano/gpuarray/subtensor.py

```diff
@@ -582,6 +582,128 @@ def perform(self, node, inputs, out_):
+        nidx = []
+        nshp = list(x.shape)
+        for k, i in enumerate(idx):
+            if i is None:
+                nidx.append(slice(None))
+                nshp.insert(k, 1)
+            else:
+                nidx.append(i)
+
+        x_ = x.reshape(nshp)
+
+        narrays = 0
+        transp = list(range(x_.ndim))
+        p = 0
+        ap = 0
+        for k, i in enumerate(list(nidx)):
+            if (isinstance(i, np.ndarray) and
+                    i.ndim != 0):
+                transp.remove(k)
+                transp.insert(p, k)
+                ap += k
+                i = nidx.pop(k)
+                nidx.insert(p, i)
+                p += 1
+                narrays += 1
+            else:
+                if narrays == 0:
+                    try:
+                        i.__index__()
+                        if ap >= 0:
+                            ap -= 1
+                        narrays = 2
+                    except Exception:
+                        pass
+        # End of copied code from AdvancedSubtensor
+        # transp: order to shuffle axes of x so that single dimension
+        #         subarrays are extracted first
+        # p: number of axes with array indexing
+
+        x_ = x_.transpose(*transp)
+        idx_ = ([slice(None)] * p + nidx[p:])
+        x_ = x_.__getitem__(idx_)
+
+        # flatten the array-indexed dimensions
+        x_flat = x_.reshape((np.prod(x_.shape[0: p]),) + x_.shape[p:])
+
+        # process y so that last axes are the same
+        if y.shape != (1,):
+            y_shape_reverse = []
+            for x_s, y_s in zip(x_flat.shape[::-1], y.shape[::-1]):
+                if x_s == y_s or y_s == 1:
+                    y_shape_reverse.append(y_s)
+                else:
+                    break
+            if np.prod(y_shape_reverse) < np.prod(y.shape):
+                if len(y_shape_reverse) > 0:
+                    y_shape_reverse.append(
+                        int(np.prod(y.shape[0:-len(y_shape_reverse)])))
+                else:
+                    y_shape_reverse.append(int(np.prod(y.shape)))
+
+            y_shape = y_shape_reverse[::-1]
+            assert(np.prod(y_shape) == np.prod(y.shape))
+            y_flat = y.reshape(y_shape)
+        else:
+            y_flat = y[0]
+
+        # build the strides
+        strides = [1]
+        for i in range(p - 1, 0, -1):
+            stride = x_.shape[i] * strides[0]
+            strides.insert(0, stride)
+
+        # build the indices and use it
+        take_idx = sum(i * s for i, s in zip(nidx, strides))
+
+        if x_flat.shape[-len(y_flat.shape):] == y_flat.shape or y_flat.shape == ():
+            # y_flat has to be broadcast over axes of x_flat[i]
+            for i in take_idx.flatten():
+                tmp = pygpu.elemwise.elemwise2(
+                    x_flat[i], '+', y_flat, x_flat[i],
+                    broadcast=True
+                )
+                x_flat.__setitem__(i, tmp)
```

Comment:
You should be able to use iadd here too.  Is there a specific reason why you use elemwise2?